File: 09-decorator/09-decorator.php
<?php

interface Shipping
{
    public function cost();
    public function description();
}

class Standard implements Shipping
{
    public function cost()
    {
        return 5;
    }
    public function description()
    {
        return "Standard shipping";
    }
}

abstract class ShippingDecorator implements Shipping
{
    public function __construct(protected Shipping $shipping)
    {
    }
    public function cost()
    {
        return $this->shipping->cost();
    }
    public function description()
    {
        return $this->shipping->description();
    }
}

class CashOnDelivery extends ShippingDecorator
{
    public function cost()
    {
        return parent::cost() + 2;
    }
    public function description()
    {
        return parent::description() . " with COD fee";
    }
}

$shipping = new Standard();
$cod = new CashOnDelivery($shipping);

echo $cod->description() . " costs " . $cod->cost();

File: 09-decorator/09-decorator.go
package main

import "fmt"

type Shipping interface {
	cost() int
	description() string
}

type Standard struct {
}

func (s *Standard) cost() int {
	return 5
}

func (s *Standard) description() string {
	return "Standard shipping"
}

type ShippingDecorator struct {
	shipping Shipping
}

func (s *ShippingDecorator) cost() int {
	return s.shipping.cost()
}

func (s *ShippingDecorator) description() string {
	return s.shipping.description()
}

type CashOnDelivery struct {
	Shipping
}

func (c *CashOnDelivery) cost() int {
	return c.Shipping.cost() + 2
}

func (c *CashOnDelivery) description() string {
	return c.Shipping.description() + " + cash on delivery fee"
}

func main() {
	s := &Standard{}
	cod := &CashOnDelivery{Shipping: s}
	fmt.Println(cod.description(), cod.cost())
}
